File: least_squares.cpp
#include <opencv2/opencv.hpp>
#include <vector>

using namespace cv;
using namespace std;

// Function to fit a quadratic polynomial surface to the background
void fitQuadraticSurface(const Mat& src, Mat& dst) {
    // Collect data points
    vector<Point3f> points;
    for (int y = 0; y < src.rows; ++y) {
        for (int x = 0; x < src.cols; ++x) {
            points.push_back(Point3f((float)x, (float)y, src.at<uchar>(y, x)));
        }
    }
    int npts = (int)points.size();

    // Fit a quadratic surface: f(x, y) = ax^2 + by^2 + cxy + dx + ey + f
    Mat A(npts, 6, CV_32F);
    Mat B(npts, 1, CV_32F);
    for (int i = 0; i < npts; i++) {
        float x = points[i].x;
        float y = points[i].y;
        float z = points[i].z;

        A.at<float>(i, 0) = x * x;
        A.at<float>(i, 1) = y * y;
        A.at<float>(i, 2) = x * y;
        A.at<float>(i, 3) = x;
        A.at<float>(i, 4) = y;
        A.at<float>(i, 5) = 1;
        B.at<float>(i, 0) = z;
    }

    Mat coeffs;
    solve(A, B, coeffs, DECOMP_SVD);

    // Create the background model
    Mat bkg = A * coeffs;
    dst = bkg.reshape(0, src.rows);
}

int main(int argc, char** argv) {
    Mat image = imread("sample_text.png", IMREAD_GRAYSCALE);
    if (image.empty()) {
        cout << "Error: Could not open or find the image!" << endl;
        return -1;
    }

    // Fit and remove the background
    Mat background, corrected, binary;
    fitQuadraticSurface(image, background);
    background.convertTo(background, CV_8U);
    subtract(background, image, corrected);
    threshold(corrected, binary, 0, 255, THRESH_BINARY | THRESH_OTSU);

    // Display results
    imshow("Original Image", image);
    imshow("Background", background);
    imshow("Corrected Image", corrected);
    imshow("Binary Image", binary);

    waitKey(0);
    return 0;
}

File: least_squares.py
import cv2
import numpy as np
import matplotlib.pyplot as plt
plt.rcParams.update({'font.size': 14})

def fit_quadratic_surface(image):
    rows, cols = image.shape
    X, Y = np.meshgrid(np.arange(cols), np.arange(rows))
    Z = image

    # Prepare the data for least squares fitting
    X_flat = X.flatten()
    Y_flat = Y.flatten()
    Z_flat = Z.flatten()

    A = np.vstack([X_flat**2, Y_flat**2, X_flat*Y_flat, X_flat, Y_flat, np.ones_like(X_flat)]).T
    B = Z_flat

    # Solve for the coefficients
    coeffs, _, _, _ = np.linalg.lstsq(A, B, rcond=None)

    # Create the background model
    background = (coeffs[0] * X**2 + coeffs[1] * Y**2 +
                  coeffs[2] * X * Y + coeffs[3] * X +
                  coeffs[4] * Y + coeffs[5])

    return background

def main(image_path):
    # Load the input image
    image = cv2.imread(image_path, cv2.IMREAD_GRAYSCALE)
    if image is None:
        print("Error: Could not open or find the image!")
        return

    # Fit and remove the background
    background = fit_quadratic_surface(image)
    residual = background.astype(np.float32) - image.astype(np.float32)
    residual = np.clip(residual, 0, 255).astype(np.uint8)
    _, binarized = cv2.threshold(residual, 0, 255, cv2.THRESH_BINARY | cv2.THRESH_OTSU)

    # Display results
    plt.figure(figsize=(6, 6))
    plt.subplot(2, 2, 1)
    plt.title('Original Image')
    plt.imshow(image, cmap='gray')
    plt.axis('off')
    plt.subplot(2, 2, 2)
    plt.title('Background')
    plt.imshow(background, cmap='gray')
    plt.axis('off')
    plt.subplot(2, 2, 3)
    plt.title('Residual Image')
    plt.imshow(residual, cmap='gray')
    plt.axis('off')
    plt.subplot(2, 2, 4)
    plt.title('Binary Image')
    plt.imshow(binarized, cmap='gray')
    plt.axis('off')
    plt.show()

if __name__ == "__main__":
    main('sample_circle.png')